improved immutabledict.merge_with and union

LucasMalor opened the pull request:
Now if `immutabledict.merge`_with have only one `immutabledict`, it returns that.
Also, `union` returns `other` if `self` is empty and `other` is an `immutabledict`.

Fixes: #13043

LucasMalor committed:
improved immutabledict.union and merge_with

Now immutabledict.union returns other if self is empty and other is an
immutabledict, and if merge_with have only one immutabledict, it returns that.

LucasMalor committed:
Merge branch 'sqlalchemy:main' into main

CaselIT commented on lib/sqlalchemy/util/_immutabledict_cy.py:
```diff
@@ -163,25 +165,46 @@ def union(
 
     @cython.annotation_typing(False)  # avoid cython crash from generic return
     def merge_with(
```
the change here doesn't seem correct

CaselIT reviewed (changes requested):
Thanks for getting the ball rolling, I've left a few comments

LucasMalor commented on lib/sqlalchemy/ext/asyncio/engine.py:
```diff
@@ -590,7 +591,8 @@ async def stream(
             execution_options=util.EMPTY_DICT.merge_with(
-                execution_options, {"stream_results": True}
+                execution_options,
+                immutabledict(stream_results=True),
```
Because if the execution_options is empty, `merge_with` can't return the other dict as-is if it's not an `immutabledict`.

Side note: the kwargs constructor is used since this way we can avoid the creation of an intermediate `dict`.

CaselIT replied:
it's kinda the same though, since we still have to construct the a new immutable dict. whatever it's not here or in the merge_with it doesn't change much

LucasMalor replied:
> we still have to construct the a new immutable dict

Previously we constructed a `dict`. Since `immutabledict` inherits from `dict`, this new creation has the same speed of `dict(k=v)` (that I suppose is fast enough). If you want, I can do some benchs.

LucasMalor replied:
> I guess at this point we could have the stream only as module variable

I missed that. What do you mean?

LucasMalor commented on lib/sqlalchemy/util/_immutabledict_cy.py:
```diff
@@ -146,11 +146,13 @@ def __repr__(self) -> str:
             return self
+
+        if not self and type(other) == immutabledict:
+            return self
```
Yes, of course :) I'll fix it.

LucasMalor committed:
fix to immutabledict.union (typo)

when self is empty and other is an immutabledict, we have to return other and not self...

LucasMalor committed:
immutabledict.merge_with accepts also None
restored Optional typing

LucasMalor committed:
minor: removed no more needed enumerate

LucasMalor committed:
test: added unit tests for modified immutabledict.merge_with and union

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision bd53e488432edd5986c28f196b0363b976b26b04 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented on lib/sqlalchemy/util/_immutabledict_cy.py:
```diff
@@ -163,25 +165,46 @@ def union(
-                    dict.update(result, d)
+                    only_one = None
+                    break
+            else:
```
**LucasMalor** (LucasMalor) wrote:

I think it's better to cycle over dicts instead of the index, it's more pythonic. If we need the index, we can use enumerate.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester reviewed (commented):
**LucasMalor** (LucasMalor) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

* lib/sqlalchemy/util/_immutabledict_cy.py (line 162): It's possible to simplify it. You need only another if and no else, since there's a return in every branch.

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

why dont we make this a little simpler and unify merge_with() and union() ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

* lib/sqlalchemy/util/_immutabledict_cy.py (line 162): Done

sqla-tester commented on lib/sqlalchemy/util/_immutabledict_cy.py:
```diff
@@ -163,25 +165,46 @@ def union(
+                    break
+            else:
+                only_one = None
+                break
```
**Michael Bayer** (zzzeek) wrote:

I think there's still a path here where we make a new dict that we dont need to (this is also a common case):

      d = util.EMPTY_DICT.merge_with(util.EMPTY_DICT, util.EMPTY_DICT)
      
we start with only_one=None, self_is_empty=True.  then into the loop through "others", we never modify "only_one" because "if not d: continue"

now we're out of the loop, only_one=None.   we make an immutabledict().   when we should be returning self.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

can you confirm my reading of the logic?   I think this creates a new dict:

   d = util.EMPTY_DICT.merge_with(util.EMPTY_DICT)
   
or if not, the logic is at least pretty hard to read then :)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

* lib/sqlalchemy/util/_immutabledict_cy.py (line 162): perfect (there's just one implementation)

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

OK, I've fixed that, let's stare at this very hard

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

Yes it wasn't token into consideration. Your impl seeems fine

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester reviewed (commented):
**LucasMalor** (LucasMalor) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester commented on lib/sqlalchemy/util/_immutabledict_cy.py:
```diff
@@ -163,25 +165,46 @@ def union(
+                break
 
-        return self if result is None else result
+        if only_one is not None:
```
**Michael Bayer** (zzzeek) wrote:

this test passes:

    @methods
    def test_only_one_non_empty_no_copy(self, method):
        d = util.immutabledict()
        e = util.immutabledict({1: 2})
        d2 = method(d, e, {})
        eq_(d2, {1: 2})
        is_(d2, e)
        


steps:

1. self resolves to to false, so only_one = False, self_is_empty = True
2. loop begins through others
3. first element in the loop is "immd" (or "e" in my example above)
4. if only_one is False and isinstance(d, immutabledict) -> True
5. only_one = d  (which in this case is "immd" / "e")
6. second element in the loop is empty plain dictionary
7. "if not d: continue" -> loop continues on next element.   
8. no more elements, out of loop, only_one == the single non-empty dictionary
9. if only_one is False -> False
10. if only_one is not None: -> True
11. only_one is returned


the "if not d:" part is what prevents us from considering any other immutable / plain dict that is empty

this same test is present in the "def test_no_change_everyone_empty" test case

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6532 has been **merged**. Congratulations! :)
